lib/src/front_end/piece_factory.dart

```diff
@@ -828,17 +831,96 @@ mixin PieceFactory {
-  Piece createList(Iterable<AstNode> elements,
+  ///
+  /// If [preserveNewlines] is `true`, then any newlines or lack of newlines
+  /// between pairs of elements in the input are preserved in the output. This
+  /// is used for collection literals that contain line comments to preserve
+  /// the author's deliberate structuring, as in:
+  ///
+  ///     matrix = [
+  ///       1, 2, 3, //
+  ///       4, 5, 6,
+  ///       7, 8, 9,
+  ///     ];
+  Piece createList(List<AstNode> elements,
       {Token? leftBracket,
       Token? rightBracket,
-      ListStyle style = const ListStyle()}) {
+      ListStyle style = const ListStyle(),
+      bool preserveNewlines = false}) {
     var builder = DelimitedListBuilder(this, style);
+
     if (leftBracket != null) builder.leftBracket(leftBracket);
-    elements.forEach(builder.visit);
+
+    if (preserveNewlines && elements.containsLineComments(rightBracket!)) {
+      _preserveNewlinesInCollection(elements, builder);
+    } else {
+      elements.forEach(builder.visit);
+    }
+
     if (rightBracket != null) builder.rightBracket(rightBracket);
     return builder.build();
   }
 
+  /// Writes [elements] into [builder], preserving the original newlines (or
+  /// lack thereof) between elements.
+  ///
+  /// This is used for formatting collection literals that contain at least one
+  /// line comment between elements. In that case, we use the line comment as a
+  /// single to prefer the author's chosen newlines between elements. For
+  /// example, if the user writes:
+  ///
+  ///     list = [
+  ///       1,2,   3, 4,
+  ///       // comment
+  ///       5,6,    7
+  ///     ];
+  ///
+  /// The formatter produces:
+  ///
+  ///     list = [
+  ///       1, 2, 3, 4,
+  ///       // comment
+  ///       5, 6, 7
+  ///     ];
+  void _preserveNewlinesInCollection(
+      List<AstNode> elements, DelimitedListBuilder builder) {
+    // Builder for all of the elements on a single line. We use a ListPiece for
+    // this too because even though we prefer to keep all elements that are on
+    // a single line in the input also on a single line in the output, we will
+    // split them if they don't fit.
+    var lineStyle = const ListStyle(commas: Commas.nonTrailing);
+    var lineBuilder = DelimitedListBuilder(this, lineStyle);
+    var atLineStart = true;
+
+    for (var i = 0; i < elements.length; i++) {
+      var element = elements[i];
+
+      if (!atLineStart &&
+          comments.hasNewlineBetween(
+              elements[i - 1].endToken, element.beginToken)) {
+        // This element begins a new line. Add the elements on the previous
+        // line to the list builder and start a new line.
+        builder.add(lineBuilder.build());
+        lineBuilder = DelimitedListBuilder(this, lineStyle);
+        atLineStart = true;
+      } else if (!atLineStart) {
+        // This element follows another on the same line.
+        // lineBuilder.space();
```

Comment:
Do we need this else if and the comment?

Reply:
Oops, you're right. That was a remnant of a different approach where I was using AdjacentPieceBuilder for the line and not DelimitedListBuilder. Removed. Thanks!